pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/refreshsegment/RefreshSegmentTaskGenerator.java

```diff
@@ -0,0 +1,171 @@
+        if (tableConfig.getTableType() == TableType.REALTIME && !segmentZKMetadata.getStatus().isCompleted()) {
+          continue;
+        }
+
+        // Skip segments for which a task is already running.
+        if (runningSegments.contains(new Segment(tableNameWithType, segmentZKMetadata.getSegmentName()))) {
+          continue;
+        }
+
+        String segmentName = segmentZKMetadata.getSegmentName();
+
+        // Skip if the segment is already up-to-date and doesn't have to be refreshed.
+        if (!shouldRefreshSegment(segmentZKMetadata, tableConfig, tableStat, schemaStat)) {
+          continue;
+        }
+
+        Map<String, String> configs = new HashMap<>(getBaseTaskConfigs(tableConfig, List.of(segmentName)));
+        configs.put(MinionConstants.DOWNLOAD_URL_KEY, segmentZKMetadata.getDownloadUrl());
+        configs.put(MinionConstants.UPLOAD_URL_KEY, _clusterInfoAccessor.getVipUrl() + "/segments");
+        configs.put(MinionConstants.ORIGINAL_SEGMENT_CRC_KEY, String.valueOf(segmentZKMetadata.getCrc()));
+        pinotTaskConfigs.add(new PinotTaskConfig(taskType, configs));
+        tableNumTasks++;
+      }
+
+      LOGGER.info("Finished generating {} tasks configs for table: {} " + "for task: {}", tableNumTasks,
+          tableNameWithType, taskType);
+    }
+
+    return pinotTaskConfigs;
+  }
+
+  /**
+   * We need not refresh when: There were no tableConfig or schema updates after the last time the segment was
+   * refreshed by this task.
+   *
+   * Note that newly created segments after the latest tableConfig/schema update will still need to be refreshed. This
+   * is because inverted index created is disabled by default during segment generation. This can be added as an
+   * additional check in the future, if required.
+   */
+  private boolean shouldRefreshSegment(SegmentZKMetadata segmentZKMetadata, TableConfig tableConfig, Stat tableStat,
+      Stat schemaStat) {
+    String tableNameWithType = tableConfig.getTableName();
+    String timestampKey = RefreshSegmentTask.TASK_TYPE + MinionConstants.TASK_TIME_SUFFIX;
+
+    long lastProcessedTime = 0L;
+    if (segmentZKMetadata.getCustomMap() != null && segmentZKMetadata.getCustomMap().containsKey(timestampKey)) {
+      lastProcessedTime = Long.parseLong(segmentZKMetadata.getCustomMap().get(timestampKey));
+    }
+
+    if (tableStat == null || schemaStat == null) {
+      LOGGER.warn("Table or schema stat is null for table: {}", tableNameWithType);
+      return false;
+    }
+
+    long tableMTime = tableStat.getMtime();
+    long schemaMTime = schemaStat.getMtime();
+
+//    TODO: See comment above - add this later if required.
+//    boolean segmentCreatedBeforeUpdate =
+//        tableMTime > segmentZKMetadata.getCreationTime() || schemaMTime > segmentZKMetadata.getCreationTime();
+
+    boolean segmentProcessedBeforeUpdate = tableMTime > lastProcessedTime || schemaMTime > lastProcessedTime;
+    return segmentProcessedBeforeUpdate;
```

Comment:
We can also add a crc check to figure out if we need to trigger a refresh or not. This way we also ensure deepstore copy gets updated with latest indexes / schemas.

Reply:
Didn't get this. Are you suggesting we need to check the crc in the ZK metadata against the crc in deepstore? They are bound to always be the same right?

Reply:
Necessarily not! I have seen this a lot of times in Upsert-compaction task as well: #13491.
This is some race-condition which we should solve fundamentally but I think for now we can let this task refresh the segment in deepstore anyways.